### PR TITLE
Allgather-based pipelining HostIR implementation

### PR DESCRIPTION
**Pull request by nsarka**
This PR has Sam's PR for preallocated buffers (https://github.com/NVIDIA/Fuser/pull/3171)

Otherwise, this is very similar to Sam's RS-based pipelining HostIR implementation

### CODE REVIEW

File: csrc/host_ir/executor.cpp [hunk 1]
```diff
@@ -281,6 +281,25 @@ void HostIrExecutor::handle(EndCoalescing* end_coalescing) {
...
+void HostIrExecutor::handle(MatmulOp* matmul) {
+  TensorView* a = matmul->inA();
+  TensorView* b = matmul->inB();
+  TensorView* out = matmul->out();
+  NVF_ERROR(
+      expr_evaluator_.isKnown(a) && expr_evaluator_.isKnown(b),
+      "Inputs of the matmul ",
+      matmul->toString(),
+      "must be precomputed before being retrieved");
+  if (expr_evaluator_.isKnown(out)) {
```
Comment (wujingyue) [inline csrc/host_ir/executor.cpp:293]: Can you verify both branches are tested? I suspect the test you added to test_host_irs only covers the if-else branch.

File: csrc/host_ir/executor.cpp [hunk 2]
```diff
@@ -281,6 +281,25 @@ void HostIrExecutor::handle(EndCoalescing* end_coalescing) {
   works_[end_coalescing] = backend->endCoalescing();
 }
 
+void HostIrExecutor::handle(MatmulOp* matmul) {
```
Comment (wujingyue) [inline csrc/host_ir/executor.cpp:284]: I recommend we move this and the unit test in test_host_irs.cpp to a different PR. https://google.github.io/eng-practices/review/developer/small-cls.html

File: tests/cpp/test_multidevice_overlap.cpp [hunk 3]
```diff
@@ -331,9 +333,8 @@ TEST_F(
   TensorView* tva_j = select(tva, 0, j);
   TensorView* tvc_j = select(tvc, 0, j);
   TensorView* tvc_locally_reduced_j =
-      matmul(tva_j, tvb); // ideally we should use the preallocated global
-                          // buffer tc_locally_reduced, but ExpressionEvaluator
-                          // do not support preallocated output buffer.
+      select(tvc_locally_reduced, 0, stream_index);
+  auto matmul = IrBuilder::create<MatmulOp>(tvc_locally_reduced_j, tva_j, tvb);
```
Comment (wujingyue) [inline tests/cpp/test_multidevice_overlap.cpp:337]: Nice! I recommend changes in ReduceScatterBasedPipeliningHostIrImplementation come as a separate PR as well. It probably has to depend on HostIrExecutor changes.

**Review by wujingyue — COMMENTED** [review]
LGTM otherwise. Thank you!

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp:293]: this comment applies to https://github.com/NVIDIA/Fuser/pull/3171
ported to https://github.com/NVIDIA/Fuser/pull/3171/files#r1800452454

File: csrc/host_ir/executor.cpp (hunk 2 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp:284]: The present PR is stacked on top of https://github.com/NVIDIA/Fuser/pull/3171 which does what you suggest

File: tests/cpp/test_multidevice_overlap.cpp (hunk 3 above)
Comment (samnordmann) [reply tests/cpp/test_multidevice_overlap.cpp:337]: same, this change is introduced in https://github.com/NVIDIA/Fuser/pull/3171

**Review by samnordmann — APPROVED** [review]
LGTM! Let's just first merge #3171 before this one

**Review by wujingyue — APPROVED** [review]

**Comment by nsarka** [comment]
!build

**Comment by nsarka** [comment]
@wujingyue It seems like I am not authorized to run blossom-ci. Can you please run it?

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
> @wujingyue It seems like I am not authorized to run blossom-ci. Can you please run it?

That's weird -- you are already in https://github.com/NVIDIA/Fuser/blob/6c3f264eeeb94a3f41844971f0ab9e7e396e112a/.github/workflows/nvfuser-ci-trigger.yml#L20, and https://github.com/nsarka is part of NVIDIA organization. 

You were able to run it before but only failed today? 

cc @xwang233

**Comment by nsarka** [comment]
!build

**Comment by samnordmann** [comment]
!build

**Comment by xwang233** [comment]
I sent a message to the support team.

**Comment by nsarka** [comment]
!build

**Comment by nsarka** [comment]
@wujingyue @xwang233 Thank you! It appears to be working now

**Pull request merged**